Improve workflow for running acceptance tests

itsjeyd opened the pull request:
This PR aims to improve the workflow for running acceptance tests in instances of the Analytics Devstack. It introduces a wrapper script that can run the whole test suite or just a specific test module/class/method. It also adds code that makes sure tests depending on services that are unavailable in a local test environment are skipped correctly.

**Test instructions**

To prepare for testing, complete the following steps:

* Test https://github.com/edx/configuration/pull/2790.

* Starting from the `analyticstack` directory, update the version of edx-analytics-pipeline that you checked out while testing https://github.com/edx/configuration/pull/2790 and run a pipeline task (make sure to adjust `--vagrant-path` before):

  ```sh
  cd ../edx-analytics-pipeline
  source venv/bin/activate
  git remote add fork https://github.com/open-craft/edx-analytics-pipeline.git
  git fetch fork
  git checkout tim/improve-pipeline-workflow
  make bootstrap
  export WHEEL_URL=https://edx-wheelhouse.s3-website-us-east-1.amazonaws.com/Ubuntu/precise
  remote-task --vagrant-path /path/to/analytics/analyticstack --remote-name devstack --override-config ${PWD}/config/devstack.cfg --wheel-url $WHEEL_URL --wait ImportEnrollmentsIntoMysql --local-scheduler --interval-end $(date +%Y-%m-%d -d "tomorrow") --n-reduce-tasks 1 
  ```

 This will cause edx-analytics-pipeline to "install itself" on your Analytics Devstack (which is necessary because edx-analytics-pipeline is currently *not* installed like any other repo when creating an instance of Analytics Devstack).

To test the changes in this PR, follow these steps:

* Starting from the `analyticstack` directory, shut down Analytics Devstack and check out another copy of edx-analytics-pipeline:

  ```sh
  cd ../analyticstack
  vagrant halt
  git clone https://github.com/open-craft/edx-analytics-pipeline.git
  cd edx-analytics-pipeline
  git checkout tim/improve-pipeline-workflow
  ```

* Modify Vagrantfile to mount the repository at `/edx/app/analytics_pipeline/analytics_pipeline`:

  ```ruby
    :analytics_pipeline => {:repo => "edx-analytics-pipeline", :local => "/edx/app/analytics_pipeline/analytics_pipeline", :owner => "hadoop"},
  ```

  **Note**: As of https://github.com/edx/configuration/commit/3f6082eedf862a2729679ba51a542b0d5d5aafe5, this step is no longer necessary.

* Bring Analytics Devstack back up and use new script to run acceptance tests:

  ```sh
  vagrant up
  cd /var/lib/analytics-tasks/devstack/repo/scripts  # This is where edx-analytics-pipeline installed itself
  ./run_acceptance.sh --test test_video --update
  ```

  As mentioned above, you can also specify that you want to limit execution to a certain test class (e.g., `test_video:VideoAcceptanceTest`) or method (e.g., `test_video:VideoAcceptanceTest.test_video_timeline`) when running tests. The `--update` option ensures that the venv in which the tests will be executed is up-to-date; you can leave it off when running another test.

* Verify that the tests that have been annotated to require specific services (such as S3) are skipped when you try to run them. For example, the following command

  ```sh
  ./run_acceptance.sh --test test_course_catalog
  ```

  should output:

  ```
  Tests the workflow for the course subjects, end to end. ... SKIP: S3 is not available

  ----------------------------------------------------------------------
  XML: nosetests.xml
  ----------------------------------------------------------------------
  Ran 1 test in 0.001s

  OK (SKIP=1)
  ```

edx-webhook commented:
Thanks for the pull request, @itsjeyd! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-analytics-pipeline&number=204

smarnach commented on edx/analytics/tasks/tests/acceptance/__init__.py:
```diff
@@ -16,6 +16,9 @@
 
+SKIP_S3_TESTS = os.getenv('SKIP_S3_TESTS')
+SKIP_S3_TESTS_REASON = 'S3 is not available'
+
```
How about exporting a decorator here instead of these constants?  Like

    def when_s3_available(function):
        return unittest.skipIf(SKIP_S3_TESTS, 'S3 is not available')(function)

The tagged tests would read a bit more succinctly and to the point:

    @when_s3_available
    def test_…

instead of

    @unittest.skipIf(SKIP_S3_TESTS, SKIP_S3_TESTS_REASON)
    def test_

smarnach commented on edx/analytics/tasks/tests/acceptance/test_enrollment_validation.py:
```diff
@@ -4,13 +4,15 @@
 import gzip
 import json
 import logging
+import os
```
Why this import?

smarnach commented on scripts/run_acceptance.sh:
```diff
@@ -0,0 +1,93 @@
+UPDATE_VENV=
+
+while [ "$1" != "" ]; do
+    case $1 in
```
You should use double quotes around `$1` here (like in the line above).

smarnach commented on scripts/run_acceptance.sh:
```diff
@@ -0,0 +1,93 @@
+    case $1 in
+        -t | --tests )
+            shift
+            TESTS=$1
```
Same here.

smarnach commented on scripts/run_acceptance.sh:
```diff
@@ -0,0 +1,93 @@
+# Prepare environment as "vagrant" user
+if [ -n "$UPDATE_VENV" ]
+then
+    sudo -Hu vagrant /bin/bash<<-PREPARE
```
`sudo` has a `-s` option meaning "run the current shell given in $SHELL (basically, the current shell) as that user", which could be used here.  (No need to change anything.  The option is handy for interactive use since it saves some typing, but you did the typing here already.)

smarnach commented:
@itsjeyd Thanks for the nice instructions!  I managed to follow all steps, and setting up the devstack went much smoother than last year. :)

Running the acceptance tests also worked fine, though the repo in `/var/lib/analytics-tasks/devstack/repo` was on the `origin/release` branch after following the instructions, so the new script wasn't there yet.  Manually adding the `open-craft` fork right there and switiching to your branch fixed this problem, and running the acceptance tests went fine, so :+1: after at least reading the few minor comments I had.

The whole setup is still a bit confusing.  Among other thing, there is definitely too big a number of checkouts of `edx-analytics-pipeline` involved, but that's kind of unrelated to this ticket.

itsjeyd commented:
@mulby @brianhw Please review.

mulby commented:
A few general thoughts:

* The intent of the JSON blob config file was to enable lots of parameters to be passed into the acceptance test framework. Is there any way we could just rely on that instead of a combination of environment variables and the JSON configuration? Perhaps we just hardcode a location for the geolocation file and in our logic, check if the file exists?
* As for the S3 disabling logic, could we detect this automatically? Not sure if boto has a way to basically say "check if I have a valid connection to S3", I suppose we could do something like list buckets?
* I would like to unify your bash script with the logic already in the Makefile. At a minimum define a new target in the Makefile and clarify the difference between the two.
* I love the decorators! Could you create one for Vertica and the exporter as well and roll that out to the tests that require them? I want to have a consistent pattern for skipping tests.

Great work! Thanks for helping us make this better for everyone.

itsjeyd commented:
Thanks @mulby for the comments!

> The intent of the JSON blob config file was to enable lots of parameters to be passed into the acceptance test framework. Is there any way we could just rely on that instead of a combination of environment variables and the JSON configuration? Perhaps we just hardcode a location for the geolocation file and in our logic, check if the file exists?

Providing a sensible default for the location of the geolocation file is a good idea I think. Where do you normally store it?

Beyond that the mechanism for specifying the path to a geolocation file on the command line might be worth keeping around as a convenience. (I thought implementing something like this was what you had in mind when you suggested to "provide a mechanism for users to download the file themselves and provide it to the test", but I might have misinterpreted :)) But let me know, I'll remove it if you think it's not useful.

> I would like to unify your bash script with the logic already in the Makefile. At a minimum define a new target in the Makefile and *clarify the difference between the two*.

Do you mean clarify the difference between a new make target for acceptance tests and the existing make target for unit tests ("test-local")? If you could elaborate a bit on what you have in mind here, that would be great.

> I love the decorators! Could you create one for Vertica and the exporter as well and roll that out to the tests that require them? I want to have a consistent pattern for skipping tests.

Thanks! Done :)

mulby commented:
> Where do you normally store it?

I haven't actually ever run them on devstack before... maybe somewhere in /var/tmp?

> Beyond that the mechanism for specifying the path to a geolocation file on the command line might be worth keeping around as a convenience. (I thought implementing something like this was what you had in mind when you suggested to "provide a mechanism for users to download the file themselves and provide it to the test", but I might have misinterpreted :)) But let me know, I'll remove it if you think it's not useful.

I'm not convinced it's worth the effort if we just document where they should put the file. I don't feel strongly about this as long as the code only has to look one place for the path (in the JSON blob), if the wrapper re-wrote the JSON before passing it into python, that would work for me too.

> Do you mean clarify the difference between a new make target for acceptance tests and the existing make target for unit tests ("test-local")? If you could elaborate a bit on what you have in mind here, that would be great.

I actually meant the "test-acceptance" target.

itsjeyd commented:
@mulby 

> I don't feel strongly about this as long as the code only has to look one place for the path (in the JSON blob), ...

OK, I removed the option for specifying the path to the geolocation file on the command line, and added a default path to the JSON blob over on https://github.com/edx/configuration/pull/2790.

> As for the S3 disabling logic, could we detect this automatically? Not sure if boto has a way to basically say "check if I have a valid connection to S3", I suppose we could do something like list buckets?

Doing something like this in `@when_s3_available` might work:

```python
def when_s3_available(function):
    s3_available = True
    import boto
    connection = boto.connect_s3()  # This will not error out if
                                    # AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY are set,
                                    # so it can't be used to check if we have a valid connection to S3
    try:
        connection.get_all_buckets()
    except boto.exception.S3ResponseError:
        s3_available = False
    return unittest.skipIf(
        not s3_available, 'S3 is not available'
    )(function)
```

With this in place, annotated tests are skipped on Analytics Devstack. Assuming that the Jenkins environment uses valid/non-dummy values for `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` (and any other relevant settings), `get_all_buckets` should not error out in that environment (allowing the tests to run). Is there any chance you could verify this on your end?

smarnach commented:
@itsjeyd Two minor comments about your proposed decorator:

1. The way it is written, it will try to connect to S3 and list all buckets for every single usage of the decorator.  It would make sense to do the test only once, and cache the result in a global variable (e.g. in `when_s3_available._have_s3`).
2. I suspect that `boto.connect_s3()` can fail, e.g. when the access key and id are not set at all (not sure).  If it can fail, it should also be enclosed in a try/except block.

itsjeyd commented:
@smarnach Thanks @smarnach for the feedback on the changes I proposed for `@when_s3_available`!

@mulby I added a new make target to the Makefile (cf. most recent commit) As for clarifying the difference between it and the existing target (test-acceptance): Were you using the existing target to run tests in a Jenkins environment? Or, more generally, what use cases does it cover that go beyond running acceptance tests locally?

mulby commented:
@itsjeyd - correct, that target is used by jenkins. However, the Jenkins job also does a bunch of stuff before it calls that target (setting up the exporter etc). My gut says we should have at least two targets, one for devstack/local and one that is more explicitly for jenkins.

itsjeyd commented:
@mulby OK thanks.

I added information about running acceptance tests to the README.

The only item from your original feedback that hasn't been resolved completely is the question of whether or not we can detect S3 availability automatically. Could you please confirm whether the check I proposed in [this comment](https://github.com/edx/edx-analytics-pipeline/pull/204#issuecomment-188212477) would pass in a Jenkins environment?

mulby commented:
@itsjeyd - verified. I did the following:

* created an empty jenkins job
* added a virtualenv build step that just runs "pip install boto==2.22.1"
* added a python step that runs
```python
s3_available = True
import boto
connection = boto.connect_s3()
try:
    connection.get_all_buckets()
except boto.exception.S3ResponseError:
    s3_available = False
print s3_available
```
* Ran the job. It printed "True" to the console.

itsjeyd commented:
@mulby Thanks! I updated `when_s3_available` to use the check (with caching).

All your original comments have now been addressed, and I also rebased both PRs for this feature (https://github.com/edx/configuration/pull/2790 and this one). Let me know if there's anything else.

mulby commented on scripts/run_acceptance.sh:
```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env bash
```
I'm nervous this file actually confuses things more than it helps. I'm trying to move away from the situation where you need two copies of the pipeline (one to manage the other) in order to run stuff.

I think I would prefer something like this:

* Assume that there is a working pipeline that has been setup with `remote-task` somewhere
* cd into that directory and activate the virtualenv
* run something like "make test-acceptance" which would, in turn, depend on `develop` and `test-requirements`
* alternately you could run "make test-acceptance-local" which would only depend on `develop-local`

Perhaps we could have a small wrapper script for the devstack that does all of the above:
```bash
cd /var/lib/analytics-tasks-devstack/repo
. ../venv/bin/activate
make test-acceptance
```

Then the rest of the logic could live in the Makefile. I'm envisioning something like this:
```
REMOTE_TASK=$(shell which remote_task) LUIGI_CONFIG_PATH='config/test.cfg' ACCEPTANCE_TEST_CONFIG="/var/tmp/acceptance.json" python -m coverage run --rcfile=./.coveragerc -m nose --nocapture --with-xunit -A acceptance $(ONLY_TESTS)
```

itsjeyd replied:
@mulby I updated the script and Makefile; the roles are now reversed. Instead of the Makefile calling the script, the script now calls `make` to run tests (and the actual test logic lives in the Makefile). To run the entire test suite, you can do `./run_acceptance.sh` from the `scripts` folder, and to run specific tests you can do `./run_acceptance.sh --tests test_course_catalog`.

I did not add `develop-local` as a dependency to `test-acceptance-local` because I think it's more convenient to be able to skip updating the virtualenv (at least for the purpose of running acceptance tests locally). This is the default behavior; if you want to force an update you can call the script with `--update`.

mulby replied:
I generally like the change. I am still wary of putting logic outside of the Makefile. It continues to be unclear to me where this would be run from.

It seems like you would run it from "/edx/app/analytics_pipeline/analytics_pipeline", but it "targets" the other pipeline installation on the box in /var/lib/. This is confusing to me and in the long term I hope to only have 1 version of the pipeline installed on the box and have everything run from it. Does that make sense?

itsjeyd replied:
@mulby The script would be run from `/var/lib/analytics-tasks-devstack/repo/scripts`. For now -- once we address [OLIVE-22](https://openedx.atlassian.net/browse/OLIVE-22) ("Make it so pipeline is installed like any other component"), it should be possible to run it from `/edx/app/analytics_pipeline/analytics_pipeline/scripts`, assuming that the pipeline will get its own venv that is equivalent to the venv for `/var/lib/analytics-tasks-devstack/repo` during provisioning.

The idea behind the current setup is the following: `/edx/app/analytics_pipeline/analytics_pipeline/` is the default `TASKS_REPO` that code is checked out from when running acceptance tests (cf. [acceptance.json](https://github.com/edx/configuration/pull/2790/files#diff-ef286cd07b3ec9b07b2dec4c950a0951)). This makes it possible to work on pipeline code locally and test it without having to push it to a remote repository first. Since `TASKS_BRANCH` is set to `HEAD` by default, the code to test can also live on any branch. The only overhead is that you might need to check out a different branch to `/var/lib/analytics-tasks-devstack/repo/` if the tests you want to run don't live on the currently checked out branch. But this is again only temporary. Once `/edx/app/analytics_pipeline/analytics_pipeline` becomes the default (and only) location for the pipeline, the test code and the code to test will automatically come from the same branch.

---

I updated the documentation to reflect the changes from https://github.com/open-craft/edx-analytics-pipeline/commit/d2f487c9ff27d4adea1b95a77e8ba8041e2be6df.

mulby replied:
OK, I think I understand now.

The only part of this that is still counter-intuitive to me is that the virtualenv for /var/lib/analytics-tasks/devstack will be used to run the code in /edx/app/analytics_pipeline/analytics_pipeline.

I think this is a good step forward, though. We can always iterate later.

mulby commented:
Generally looking great. Just one concern about the Makefile/shell script.

mulby commented on scripts/run_acceptance.sh:
```diff
@@ -0,0 +1,84 @@
+TASKS_REPO=/edx/app/analytics_pipeline/analytics_pipeline
+
+# Check if repo has been cloned and mounted at default location
+if [ ! -d "$TASKS_REPO" ]
```
Is this check necessary? (also the very existence of the TASK_REPO variable)

itsjeyd replied:
@mulby For the setup described [above](https://github.com/edx/edx-analytics-pipeline/pull/204#discussion_r55066739), yes.

mulby replied:
Oh, I understand now. Thanks for the clarification.

mulby commented on scripts/run_acceptance.sh:
```diff
@@ -0,0 +1,84 @@
+
+BASE_DIRECTORY=/var/lib/analytics-tasks/devstack
+VENV_DIRECTORY=$BASE_DIRECTORY/venv
+WORKING_DIRECTORY=$BASE_DIRECTORY/repo
```
Are all of these variables still used?

mulby commented on scripts/run_acceptance.sh:
```diff
@@ -0,0 +1,84 @@
+
+# Tests
+
+if [ -n "$TESTS" ]
```
This works differently than ONLY_TESTS which I suspect could be confusing. The name also implies that you can specify multiple tests but the logic seems to imply that you should only be specifying one.

mulby replied:
I should clarify, since ONLY_TESTS is passed directly in to nose you can actually specify multiple tests and there are a bunch of different formats you can use to "choose" which tests get run.

mulby commented on scripts/run_acceptance.sh:
```diff
@@ -0,0 +1,84 @@
+
+# Export AWS keys
+export AWS_ACCESS_KEY_ID='AK123'
+export AWS_SECRET_ACCESS_KEY='abc123'
```
Why AWS keys here?

mulby commented on Makefile:
```diff
@@ -49,6 +49,9 @@ test: test-requirements develop test-local
 	LUIGI_CONFIG_PATH='config/test.cfg' python -m coverage run --rcfile=./.coveragerc -m nose --nocapture --with-xunit -A acceptance $(ONLY_TESTS)
 
+test-acceptance-local:
+	AWS_ACCESS_KEY_ID='AK123' AWS_SECRET_ACCESS_KEY='abc123' REMOTE_TASK=$(shell which remote-task) LUIGI_CONFIG_PATH='config/test.cfg' ACCEPTANCE_TEST_CONFIG="/var/tmp/acceptance.json" python -m coverage run --rcfile=./.coveragerc -m nose --nocapture --with-xunit -A acceptance --stop -v $(ONLY_TESTS)
```
heh - sorry to keep asking about this... I still don't understand why we define AWS access keys here?

itsjeyd replied:
@mulby When I started working on this, a few tests (that turned out not to depend on S3) were failing with something similar to this:

```
Traceback (most recent call last):
  File "/var/lib/analytics-tasks/devstack/venv/local/lib/python2.7/site-packages/boto/utils.py", line 219, in retry_url
    r = opener.open(req)
  File "/usr/lib/python2.7/urllib2.py", line 431, in open
    response = self._open(req, data)
  File "/usr/lib/python2.7/urllib2.py", line 449, in _open
    '_open', req)
  File "/usr/lib/python2.7/urllib2.py", line 409, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 1227, in http_open
    return self.do_open(httplib.HTTPConnection, req)
  File "/usr/lib/python2.7/urllib2.py", line 1197, in do_open
    raise URLError(err)
URLError: <urlopen error timed out>
boto: ERROR: Unable to read instance data, giving up
```

Setting `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` before running these tests fixed the problem. However, as of today I'm unable to reproduce the problem, so I removed the definitions from the Makefile (cf. https://github.com/open-craft/edx-analytics-pipeline/commit/5857e648ed516638bb42ea6c9eaff1347b93df43).

mulby commented:
@brianhw can you give this a quick review as well?

mulby commented on share/task.yml:
```diff
@@ -82,7 +83,7 @@
 
     - name: analytics tasks repository checked out
-      git: repo={{ repo }} dest={{ working_repo_dir }} version=release
+      git: repo={{ repo }} dest={{ working_repo_dir }} version={{ remote }}/{{ branch }}
```
Is this change necessary? it would require quite a bit of rework to our Jenkins jobs...

itsjeyd replied:
@mulby Yes, please see the commit message of https://github.com/open-craft/edx-analytics-pipeline/commit/f6a2ca5d38ec72cc2129a652db356a572c43c3d1 for the reasoning.

mulby replied:
What if we just checked out a bare copy of the repo? I think that would solve both of our issues... it would remove the local copy of the release branch and still allow me to pass in "origin/foo" for the branch parameter.

itsjeyd replied:
@mulby I made a mistake yesterday when testing the approach you suggested (forgot to update the venv, so modifications from https://github.com/open-craft/edx-analytics-pipeline/commit/33537a51a29ec6d4f90cb89d1630d80087a7bbdb weren't being used). It turns out that when replacing `version=release` with `bare=yes`, the playbook fails because the `branch checked out` task needs a work tree to succeed:

```
TASK: [branch checked out] **************************************************** 
<localhost> ESTABLISH CONNECTION FOR USER: hadoop
<localhost> EXEC ['ssh', '-tt', '-vvv', '-o', 'ForwardAgent=yes', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'StrictHostKeyChecking=no', '-o', 'Port=22', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'ConnectTimeout=10', 'localhost', "/bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-1457430542.22-182101216527929 && chmod a+rx $HOME/.ansible/tmp/ansible-1457430542.22-182101216527929 && echo $HOME/.ansible/tmp/ansible-1457430542.22-182101216527929'"]
<localhost> REMOTE_MODULE command git checkout HEAD chdir=/var/lib/analytics-tasks/2c0a5e8921aa235b148da415269e91c9/repo
<localhost> PUT /tmp/tmpHN8Fdg TO /edx/app/hadoop/.ansible/tmp/ansible-1457430542.22-182101216527929/command
<localhost> EXEC ['ssh', '-tt', '-vvv', '-o', 'ForwardAgent=yes', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'StrictHostKeyChecking=no', '-o', 'Port=22', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'ConnectTimeout=10', 'localhost', "/bin/sh -c '/usr/bin/python /edx/app/hadoop/.ansible/tmp/ansible-1457430542.22-182101216527929/command; rm -rf /edx/app/hadoop/.ansible/tmp/ansible-1457430542.22-182101216527929/ >/dev/null 2>&1'"]
failed: [localhost] => {"changed": true, "cmd": ["git", "checkout", "HEAD"], "delta": "0:00:00.001818", "end": "2016-03-08 09:49:03.546037", "rc": 128, "start": "2016-03-08 09:49:03.544219"}
stderr: fatal: This operation must be run in a work tree

FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/edx/app/hadoop/task.retry

localhost                  : ok=8    changed=4    unreachable=0    failed=1   

Exiting with status = 2
```

I reverted the changes. If you have any other suggestions that would help avoid making changes to Jenkins jobs, please let me know. Otherwise this should be good to go (addressed Brian's pointers about the documentation and squashed the commits).

mulby replied:
In that case, can we just revert these changes entirely?

I don't see any problem with having a ghost release branch sitting on the machine, and it will cause a huge amount of work on our end to go through all of our jenkins jobs and fix this.

Once we have the jobs checked in as DSL files, this type of change will be much easier to make :smile:

itsjeyd replied:
@mulby 

> In that case, can we just revert these changes entirely?

That would be possible, at the cost of adding another manual setup step: `/edx/app/analytics_pipeline/analytics_pipeline` needs to have a local tracking branch for `origin/release`, otherwise the `analytics tasks repository checked out` task will fail:

```
TASK: [analytics tasks repository checked out] ******************************** 
<localhost> ESTABLISH CONNECTION FOR USER: hadoop
<localhost> EXEC ['ssh', '-tt', '-vvv', '-o', 'ForwardAgent=yes', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'StrictHostKeyChecking=no', '-o', 'Port=22', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'ConnectTimeout=10', 'localhost', "/bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-1457461780.75-16575766786959 && chmod a+rx $HOME/.ansible/tmp/ansible-1457461780.75-16575766786959 && echo $HOME/.ansible/tmp/ansible-1457461780.75-16575766786959'"]
<localhost> REMOTE_MODULE git repo=/edx/app/analytics_pipeline/analytics_pipeline dest=/var/lib/analytics-tasks/2c0a5e8921aa235b148da415269e91c9/repo version=release
<localhost> PUT /tmp/tmpWy3sm0 TO /edx/app/hadoop/.ansible/tmp/ansible-1457461780.75-16575766786959/git
<localhost> EXEC ['ssh', '-tt', '-vvv', '-o', 'ForwardAgent=yes', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'StrictHostKeyChecking=no', '-o', 'Port=22', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'ConnectTimeout=10', 'localhost', "/bin/sh -c '/usr/bin/python /edx/app/hadoop/.ansible/tmp/ansible-1457461780.75-16575766786959/git; rm -rf /edx/app/hadoop/.ansible/tmp/ansible-1457461780.75-16575766786959/ >/dev/null 2>&1'"]
failed: [localhost] => {"failed": true}
msg: Failed to checkout release

FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/edx/app/hadoop/task.retry

localhost                  : ok=6    changed=2    unreachable=0    failed=1   

Exiting with status = 2
```

Also, the default for `tasks_branch` in `acceptance.json` needs to be adjusted for this to work (`"HEAD"` ⟶ `"origin/HEAD"`).

Let me know if that's an acceptable trade-off for you; I'll make the necessary adjustments to this PR and https://github.com/edx/configuration/pull/2790 in that case.

(Automating the step of creating a local tracking branch for `origin/release` could be addressed in the context of [OLIVE-22](https://openedx.atlassian.net/browse/OLIVE-22), so the trade-off would likely be temporary.)

mulby replied:
What if we just change the default branch to "origin/release"? And then procede with the checkout normally?

Not sure I understand why we have to change HEAD to origin/HEAD but I think I'd rather do that then change this code.

itsjeyd replied:
@mulby 

> What if we just change the default branch to "origin/release"?

That doesn't make a difference (i.e., it produces the same error if `/edx/app/analytics_pipeline/analytics_pipeline` does not have a local `release` branch). This is because the copy of the pipeline that is checked out for the purpose of running tests doesn't "see" the remote tracking branches that `/edx/app/analytics_pipeline/analytics_pipeline` has. It only sees the local branches. To give a concrete example, a branch `foo` that is local to `/edx/app/analytics_pipeline/analytics_pipeline` is available to the working copy as `origin/foo`. (For the `git` module, `version=origin/release` is the same as `version=release` because the `remote` parameter -- which we're not customizing in the `analytics tasks repository checked out` task -- defaults to `origin`.)

> Not sure I understand why we have to change HEAD to origin/HEAD ...

If the version that the `analytics tasks repository checked out` task checks out (`release`) is different from the version that we want to run our tests on (the development branch that `/edx/app/analytics_pipeline/analytics_pipeline` is on), checking out `HEAD` in the `branch checked out` task that follows won't give the desired result (i.e., the current branch will stay on `release`).

---

To sum up, the following changes would make it possible for you to avoid having to modify Jenkins jobs:

1. Revert https://github.com/open-craft/edx-analytics-pipeline/commit/f6a2ca5d38ec72cc2129a652db356a572c43c3d1.
2. Update README to make it clear that `/edx/app/analytics_pipeline/analytics_pipeline` needs a local `release` branch.
3. Change value of `tasks_branch` to `origin/HEAD` in acceptance.json.

Can I go ahead with these (and merge afterwards)?

mulby replied:
I think I now understand. I was thinking of "origin" as github, but it's actually the repo stored at `/edx/app/analytics_pipeline` apologies for the confusion.

What if we changed the default branch to `master`? Since that's the default branch I think it's likely that it will exist locally in origin.

We should definitely change `tasks_branch` to `origin/HEAD`.

OK with merging once we have this resolved.

itsjeyd replied:
@mulby Changing the default branch to `master` works :) I did that instead of reverting https://github.com/open-craft/edx-analytics-pipeline/commit/f6a2ca5d38ec72cc2129a652db356a572c43c3d1 and updating the README. Adjusted `tasks_branch` in acceptance.json as well and squashed the commits for https://github.com/edx/configuration/pull/2790. The build for that PR is green. Good to go?

brianhw commented on README.md:
```diff
@@ -65,6 +68,111 @@ valid credentials for the tests to pass, so the commands below should fix the fa
+
+- `/var/lib/acceptance.json` is present on your instance of Analytics
+  Devstack. This file contains the default configuration for running
+  acceptance tests.
```
Having this documentation is great!  Is it possible to offer some guidance here as to where a user might find a template or example for this file?

itsjeyd replied:
@brianhw I added a link to the default version of acceptance.json to the README. The file is being introduced in the companion PR (https://github.com/edx/configuration/pull/2790) for this, so the link will work once that PR is merged.

brianhw commented:
One optional doc comment, otherwise this LGTM. :+1:

mulby commented:
:+1: from me as well. Ready to squash and merge :)

itsjeyd commented:
@mulby @brianhw Thanks for reviewing! Could you please also have a look at/add thumbs-up to the companion PR (https://github.com/edx/configuration/pull/2790) for this?

If possible, I'd like to merge the PRs at the same time.

brianhw commented on README.md:
```diff
@@ -65,6 +68,111 @@ valid credentials for the tests to pass, so the commands below should fix the fa
+Devstack, the following conditions must hold:
+
+- [`/var/lib/acceptance.json`](https://github.com/edx/configuration/blob/master/playbooks/roles/analytics_pipeline/files/acceptance.json)
+  is present on your instance of Analytics Devstack. This file
```
Thanks for adding the pointer.  I discovered that when reviewing the other PR.  

But I just noticied that this recommends putting the file in /var/lib.  However, the other PR puts it in /var/tmp, and the makefile expects it in /var/tmp as well.

itsjeyd replied:
@brianhw Thanks for catching that! Fixed.

brianhw commented:
Just the one correction in the documentation, then :+1:

mulby commented:
@itsjeyd ready to merge once I run the acceptance tests on this branch on our jenkins box, which I'm kicking off now

itsjeyd commented:
@mulby OK thanks, let me know.

mulby commented:
@itsjeyd: The test run passed, this is ready to merge whenever you're ready.

Great work!

bradenmacdonald commented:
@itsjeyd has either just left for vacation or is about to, so I'm going to merge this for him.

Thanks for the great work on this @itsjeyd, and thanks for the review @mulby @brianhw @smarnach.

:tada:

itsjeyd commented:
@mulby @bradenmacdonald Thanks guys!
